hub/hub.go

```diff
@@ -232,114 +223,175 @@ func (h *ForkableHub) SourceThroughCursor(startBlock uint64, cursor *bstream.Cur
+		}
 
-	if !h.forkable.Linkable(blk) {
-		fdb_head := h.forkable.HeadNum()
-		if blk.Number < fdb_head {
-			zlog.Info("live block not linkable yet, will retry when we reach forkDB's HEAD", zap.Stringer("blk_from_live", blk.AsRef()), zap.Uint64("forkdb_head_num", fdb_head))
-			return nil
+		oneBlocksUpToLibRef = append(oneBlocksUpToLibRef, currentBlock)
+
+		err = h.forkable.ProcessBlock(currentBlock, nil)
+		if err != nil {
+			return fmt.Errorf("processing block: %w", err)
 		}
-		zlog.Warn("cannot initialize forkDB from one-block-files (hole between live and one-block-files). Will retry on every incoming live block.", zap.Uint64("forkdb_head_block", fdb_head), zap.Stringer("blk_from_live", blk.AsRef()))
-		return nil
 	}
-	zlog.Info("hub is now Ready")
 
-	h.ready = true
-	close(h.Ready)
+	if !h.forkable.Linkable(oneBlocksUpToLibRef[len(oneBlocksUpToLibRef)-1]) {
+		return fmt.Errorf("most recent one block is not linkable")
+	}
+
 	return nil
 }
 
 func (h *ForkableHub) Run() {
-	liveSource := h.liveSourceFactory(bstream.HandlerFunc(h.bootstrapperHandler))
+	liveSource := h.liveSourceFactory(h)
 	liveSource.OnTerminating(h.reconnect)
+
+	err := h.bootstrap()
+	if err != nil {
+		zlog.Warn("bootstrapping incomplete", zap.Error(err))
+	} else {
+		close(h.Ready)
+	}
+
 	liveSource.Run()
+
+	zlog.Info("Hub is ready")
+
 }
+func (h *ForkableHub) ProcessBlock(blk *pbbstream.Block, obj interface{}) error {
+	ctx := context.Background()
 
-func (h *ForkableHub) reconnect(err error) {
-	failFunc := func() {
-		h.Shutdown(fmt.Errorf("cannot link new blocks to chain after a reconnection"))
-	}
+	lastKnownLib := h.forkable.LowestBlockNum()
 
-	rh := newReconnectionHandler(
-		h.forkable,
-		h.forkable.HeadNum,
-		time.Minute,
-		failFunc,
-	)
+	if !h.forkable.Linkable(blk) {
+		sortedOneBlocksFiles, err := h.WalkOneBlocksStore(ctx)
+		if err != nil {
+			return fmt.Errorf("walking through one blocks files: %w", err)
+		}
 
-	zlog.Info("reconnecting hub after disconnection. expecting to reconnect and get blocks linking to headnum within delay",
-		zap.Duration("delay", rh.timeout),
-		zap.Uint64("current_head_block_num", rh.previousHeadBlock),
-		zap.Error(err))
+		if len(sortedOneBlocksFiles) == 0 {
+			return fmt.Errorf("no one blocks found")
+		}
+
+		for _, filename := range sortedOneBlocksFiles {
+			blockNum, suffixID, _, _, _, err := bstream.ParseFilename(filename)
+			if err != nil {
+				return fmt.Errorf("parsing filename: %w", err)
+			}
+
+			if blockNum < lastKnownLib {
+				continue
+			}
+			if availableBlock := h.forkable.GetBlockByHashSuffix(suffixID); availableBlock != nil {
+				//Block already known by the forkable
+				continue
+			}
+
+			blockFromFile, err := decodeOneBlockFromFilename(ctx, filename, h.oneBlocksStore)
+			if err != nil {
+				return fmt.Errorf("decoding %s from block store: %w", filename, err)
+			}
+
+			if blockFromFile.Number == blk.LibNum {
```

Comment:
this condition should not be checked if we have no LIB on the forkdb (see prev comment)